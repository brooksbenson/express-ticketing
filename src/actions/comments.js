import db from '../firebase/firebase';

export const addComment = ({ key, ...comment }) => ({
  type: 'ADD_COMMENT',
  key,
  comment
});

export const startAddComment = text => {
  return async (dispatch, getState) => {
    const { activeUserKey: userKey, activeTicketKey } = getState();
    const comment = { date: Date.now(), userKey, text };
    const { key } = db.ref(`comments/${activeTicketKey}`).push(comment);
    dispatch(addComment({ key, ...comment }));
  };
};

export const setComments = comments => ({
  type: 'SET_COMMENTS',
  comments
});

export const startSetComments = () => {
  return async (dispatch, getState) => {
    const { activeTicketKey } = getState();
    const snap = await db.ref(`comments/${activeTicketKey}`).once('value');
    dispatch(setComments(snap.val()));
  };
};

export const unsetComments = () => ({
  type: 'UNSET_COMMENTS'
});
