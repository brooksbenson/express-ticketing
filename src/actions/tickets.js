import db from '../firebase/firebase';

export const addTicket = ({ key, ...ticket }) => ({
  type: 'ADD_TICKET',
  key,
  ticket
});

export const startAddTicket = ({ title, urgency }) => {
  return async (dispatch, getState) => {
    const state = getState();
    const ticket = {
      date: Date.now(),
      title,
      urgency,
      accountKey: state.activeAccountKey,
      contactKey: state.activeContactKey,
      userKeys: { [state.activeUserKey]: true }
    };
    const { key } = await db.ref('tickets').push(ticket);
    await Promise.all([
      db.ref('open_tickets').update({ [key]: true }),
      db.ref(`user_tickets/${state.activeUserKey}`).update({ [key]: true })
    ]);
    dispatch(addTicket({ key, ...ticket }));
    return key;
  };
};

export const setTickets = tickets => ({
  type: 'SET_TICKETS',
  tickets
});

export const startSetTickets = () => {
  return async (dispatch, getState) => {
    const state = getState();
    const [openTickets, userTickets] = await Promise.all([
      db.ref('open_tickets').once('value'),
      db.ref(`user_tickets/${state.activeUserKey}`).once('value')
    ]);
    const toFetch = [];
    const userTicketsObj = userTickets.val();
    Object.keys(openTickets.val()).forEach(key => {
      if (key in userTicketsObj) toFetch.push(key);
    });
    const tickets = {};
    await Promise.all(
      toFetch.map(
        key =>
          new Promise(resolve => {
            db.ref(`tickets/${key}`)
              .once('value')
              .then(snap => {
                tickets[key] = snap.val();
                resolve();
              });
          })
      )
    );
    dispatch(setTickets(tickets));
  };
};
