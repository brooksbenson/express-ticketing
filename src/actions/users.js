import db, { auth } from '../firebase/firebase';

export const addUser = ({ key, ...user }) => ({
  type: 'ADD_USER',
  key,
  user
});

export const startAddUser = user => {
  return async dispatch => {
    const { key } = await db.ref('user_data').push(user);
    dispatch(addUser({ key, ...user }));
    return key;
  };
};

export const setUsers = users => ({
  type: 'SET_USERS',
  users
});

export const startSetUsers = () => {
  return async dispatch => {
    const snap = await db.ref('user_data').once('value');
    dispatch(setUsers(snap.val()));
  };
};
