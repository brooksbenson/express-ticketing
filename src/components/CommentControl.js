import React from 'react';

export default props => (
  <section className={props.className}>
    <h3> Comments </h3>
    <textarea
      onChange={e => props.onCommentChange(e.target.value)}
      value={props.comment}
    />
    <button onClick={props.onCommentSave}> Save Comment </button>
    <ul>
      {props.comments.map(c => (
        <li key={c.key}>
          <div>
            <span>{c.name}</span>
            <span>{c.date}</span>
          </div>
          <p>{c.text}</p>
        </li>
      ))}
    </ul>
  </section>
);
