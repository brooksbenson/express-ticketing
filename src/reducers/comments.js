export default (state = null, action) => {
  switch (action.type) {
    case 'ADD_COMMENT':
      return { ...state, [action.key]: action.comment };
    case 'SET_COMMENTS':
      return action.comments;
    case 'UNSET_COMMENTS':
      return null;
    default:
      return state;
  }
};
