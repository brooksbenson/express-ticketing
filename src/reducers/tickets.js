export default (state = [], action) => {
  switch (action.type) {
    case 'ADD_TICKET':
      return { ...state, [action.key]: action.ticket };
    case 'SET_TICKETS':
      return action.tickets;
    case 'UPDATE_URGENCY': {
      const { key, urgency } = action;
      const ticket = state[key];
      return { ...state, [key]: { ...ticket, urgency } };
    }
    case 'ADD_USER': {
      const { key, userKey } = action;
      const ticket = state[key];
      const userKeys = { ...ticket.userKeys, [userKey]: true };
      return { ...state, [key]: { ...ticket, userKeys } };
    }
    default:
      return state;
  }
};
