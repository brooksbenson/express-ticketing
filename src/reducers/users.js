export default (state = {}, action) => {
  switch (action.type) {
    case 'ADD_USER':
      return { ...state, [action.key]: action.user };
    case 'SET_USERS':
      return action.users;
    default:
      return state;
  }
};
