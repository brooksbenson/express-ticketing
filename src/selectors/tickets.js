export default (tickets, query) => {
  const regex = new RegExp(query.split('').join('.*?'), 'i');
  return tickets.filter(
    t =>
      regex.test(t.accountName) || regex.test(t.urgency) || regex.test(t.title)
  );
};
