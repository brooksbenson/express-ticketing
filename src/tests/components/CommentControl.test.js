import React from 'react';
import CommentControl from '../../components/CommentControl';
import { shallow } from 'enzyme';
import { ticketKeys } from '../fixtures/tickets';
import { comments } from '../fixtures/comments';
import { usersObj } from '../fixtures/users';

/*
  The comments prop should be an array where
  each value is an object that has the following
  props: name, date, and text.
*/

const ticketComments = comments[ticketKeys[0]];
const commentsArr = [];

Object.keys(ticketComments).forEach(key => {
  const { userKey, ...rest } = ticketComments[key];
  commentsArr.push({ key, name: usersObj[userKey].name, ...rest });
});

const className = 'comment-control';
let wrapper;
let onCommentChange;
let onCommentSave;

beforeEach(() => {
  onCommentChange = jest.fn();
  onCommentSave = jest.fn();
  wrapper = shallow(
    <CommentControl
      className={className}
      comments={commentsArr}
      comment=""
      onCommentChange={onCommentChange}
      onCommentSave={onCommentSave}
    />
  );
});

test('CommentControl should match snapshot', () => {
  expect(wrapper).toMatchSnapshot();
});

test('CommentControl wrapper element should have the className passed as a prop', () => {
  expect(wrapper.hasClass(className)).toBeTruthy();
});

test('CommentControl should render the correct number of comments', () => {
  const { length: numberOfComments } = wrapper.find('li');
  expect(numberOfComments).toBe(commentsArr.length);
});

test('CommentControl should render a comment correctly', () => {
  const comment = wrapper.find('li').at(0);
  expect(
    comment
      .find('span')
      .at(0)
      .text()
  ).toBe(commentsArr[0].name);
  expect(
    comment
      .find('span')
      .at(1)
      .text()
  ).toBe(String(commentsArr[0].date));
  expect(
    comment
      .find('p')
      .at(0)
      .text()
  ).toBe(commentsArr[0].text);
});

test('CommentControl should respond correctly textarea change', () => {
  const value = 'Wow. This is bizarre.';
  wrapper.find('textarea').simulate('change', { target: { value } });
  expect(onCommentChange).toHaveBeenCalledWith(value);
});

test('CommentControl should respond to save comment button click correctly', () => {
  wrapper.find('button').simulate('click');
  expect(onCommentSave).toHaveBeenCalled();
});
