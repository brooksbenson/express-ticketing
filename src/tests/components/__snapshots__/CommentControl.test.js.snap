// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentControl should match snapshot 1`] = `
<section
  className="comment-control"
>
  <h3>
     Comments 
  </h3>
  <textarea
    onChange={[Function]}
    value=""
  />
  <button
    onClick={[MockFunction]}
  >
     Save Comment 
  </button>
  <ul>
    <li
      key="12233"
    >
      <div>
        <span>
          Max Benson
        </span>
        <span>
          0
        </span>
      </div>
      <p>
        Wow. This is crazy.
      </p>
    </li>
    <li
      key="14283"
    >
      <div>
        <span>
          Max Benson
        </span>
        <span>
          0
        </span>
      </div>
      <p>
        Wow. This is nuts.
      </p>
    </li>
    <li
      key="24233"
    >
      <div>
        <span>
          Max Benson
        </span>
        <span>
          0
        </span>
      </div>
      <p>
        Wow. This is insane.
      </p>
    </li>
  </ul>
</section>
`;
